Use the locale switcher to set the user language

fritzmg opened the pull request:
While I could not reproduce #8690 I think it makes sense to make sure that the `contao_default` translations are loaded into `$GLOBALS['TL_LANG']` within `AbstractLayoutPageController` (just like `PageRegular` does).

I could however reproduce this @lukasbableck

> Root page language is set to de, but everything is in English.

This PR fixes that by setting the correct language, just as `PageRegular` does. Though I was thinking, may be this should actually be done earlier by the routing? @aschempp 

In addition this PR makes two further changes:

* Add support for `$GLOBALS['TL_LANGUAGE']`. We still have modules for which there are no modern replacements that rely on it.
* Only create the response context, if none exists. In theory this should never be the case when using the `AbstractLayoutPageController`, but I think it makes sense to check for it anyway.

fritzmg committed:
Load default translation domain

fritzmg committed:
Set the current language

fritzmg committed:
update test

fritzmg committed:
add comment

Toflar commented on core-bundle/src/Controller/Page/AbstractLayoutPageController.php:
```diff
@@ -44,6 +46,17 @@ public function __invoke(Request $request): Response
+        // Override the locale based on the page's language
+        $locale = LocaleUtil::formatAsLocale($page->language);
+        $this->container->get('translator')->setLocale($locale);
+        $request->setLocale($locale);
```
Shouldn't we use the LocaleSwitcher?🧐

ausi replied:
I think changing the locale of the request is correct here as every listener that follows should also have the locale of the page. 
But as @fritzmg wrote in the initial comment, I think 
> …this should actually be done earlier by the routing…

fritzmg replied:
Yes, I agree with @ausi, we want to _define_ the locale, not switch it.

leofeyer committed:
Merge branch '5.6' into bugfix/load-default-translations-layout-page-controller

fritzmg committed:
Merge branch '5.6' into bugfix/load-default-translations-layout-page-controller

fritzmg committed:
fix test

fritzmg committed:
fix test

aschempp reviewed (changes requested):
Requesting changes because this really should already be done by routing. We have a `_locale` attribute that should be loaded by route listeners. We need to investigate why this is not working (if it is not working).

fritzmg committed:
Merge branch '5.6' into bugfix/load-default-translations-layout-page-controller

zoglo committed:
Restore the previous `autoFocus` when navigating with Turbo Drive (see #8715)

Description
-----------

### Description

- Fixes #8714

_Basically:_
The controller has been added in Contao `5.3`, pre-turbo and `this.autoFocus` was necessary to handle multiple autoFocusTargets as only the first one should be focused on.

Introcuding Turbo in `5.4` also introduced this bug (that did go unnoticed for over a year) as `this.autoFocus` has never been reset to false anymore.

The reason for not being reset whilst navigating is due to the `scroll-offset-controller` being applied to the root of the page (html). Turbo drive replaces the body, thus never affecting the controller anymore / not being able to disconnect the controller.
Setting it to false within a targetDisconnect should fix this behavior.

Commits
-------

0ae3dd25 Reset the previous `autoFocus` when the target disconnected
4728a548 Do not use a stimulus value

leofeyer committed:
Rebuild the assets

fritzmg committed:
Fix the search indexer migration when migrating from Contao 4.13 to 5.6 (see #8707)

Description
-----------

When updating from Contao 4.13 to 5.6 the following errors will occur during the database migration:

```
  [PDOException (22007)]
  SQLSTATE[22007]: Invalid datetime format: 1292 Truncated incorrect DECIMAL value: ''  


Exception trace:
  at vendor\doctrine\dbal\src\Driver\PDO\Connection.php:33
 PDO->exec() at vendor\doctrine\dbal\src\Driver\PDO\Connection.php:33
 Doctrine\DBAL\Driver\PDO\Connection->exec() at vendor\doctrine\dbal\src\Connection.php:1216
 Doctrine\DBAL\Connection->executeStatement() at vendor\contao\core-bundle\src\Migration\Version505\OverwriteMetaMigration.php:72
```
```
 [PDOException (22007)]
  SQLSTATE[22007]: Invalid datetime format: 1292 Truncated incorrect DECIMAL value: ''  


Exception trace:
  at vendor\doctrine\dbal\src\Driver\PDO\Connection.php:33
 PDO->exec() at vendor\doctrine\dbal\src\Driver\PDO\Connection.php:33
 Doctrine\DBAL\Driver\PDO\Connection->exec() at vendor\doctrine\dbal\src\Connection.php:1216
 Doctrine\DBAL\Connection->executeStatement() at vendor\contao\core-bundle\src\Migration\Version506\SearchIndexerSettingsMigration.php:46
```

The first error occurs because the `OverwriteMetaMigration` happens to be executed _before_ the `BooleanFieldsMigration`. Thus the `overwriteMeta = 1` query results in the (cryptic) SQL error, since `overwriteMeta` will still be a `CHAR(1)` at this point.

The second error occurs because we are using `noSearch = 1` - again because `noSearch` will still be a `CHAR(1)` at this point - but it will never be migrated to a `TINYINT` in this case, since it does not exist in the DCA anymore.

This PR fixes the former by making sure the `BooleanFieldsMigration` runs first and the latter by using `noSearch = '1'` - which is also safe for `TINYINT`.

Commits
-------

44dbfad1 Fix 4.13 to 5.6 migration errors
5d406f7c add comments
b3a0fd5e revert changes to the OverwriteMetaMigration (see #8708)

leofeyer committed:
Fix the PR workflow

fritzmg committed:
Modernize legacy TL_LANGUAGE super global

fritzmg committed:
Merge remote-tracking branch 'origin/5.6' into bugfix/load-default-translations-layout-page-controller

fritzmg committed:
revert changes to the LocaleSubscriber

fritzmg committed:
update tests

fritzmg committed:
Still set locale of back end request§

fritzmg committed:
update tests

fritzmg committed:
Only set global when legacy framework is initialized

fritzmg committed:
CS

fritzmg committed:
CS

Toflar commented on core-bundle/src/Framework/ContaoFramework.php:
```diff
@@ -148,6 +149,19 @@ public static function getNonce(): string
         return self::$nonce;
     }
 
+    public function setLocale(string $locale)
```
```suggestion
    public function setLocale(string $locale): void
```

fritzmg committed:
Update core-bundle/src/Framework/ContaoFramework.php

Co-authored-by: Yanick Witschi <yanick.witschi@terminal42.ch>

aschempp commented on core-bundle/src/Controller/Page/AbstractLayoutPageController.php:
```diff
@@ -44,6 +45,9 @@ public function __invoke(Request $request): Response
         }
 
+        // Load contao_default translations (#8690)
+        $this->getContaoAdapter(System::class)->loadLanguageFile('default', $request->getLocale());
```
I don't think we should pass the request locale here, it should use the current translation locale (that should already be _known_ to the framework.

aschempp commented on core-bundle/src/Controller/Page/AbstractLayoutPageController.php:
```diff
@@ -75,6 +79,10 @@ public static function getSubscribedServices(): array
 
     protected function getResponseContext(PageModel $page): ResponseContext
     {
+        if ($responseContext = $this->container->get('contao.routing.response_context_accessor')->getResponseContext()) {
```
this seems to be unrelated to ehe PR title?

fritzmg replied:
I can move that to a  separate PR, if preferred.

leofeyer replied:
Yes please. Unrelated changes in a PR make them much harder to track.

fritzmg replied:
Done 👍

aschempp commented on core-bundle/src/Framework/ContaoFramework.php:
```diff
@@ -148,6 +149,19 @@ public static function getNonce(): string
         return self::$nonce;
     }
 
+    public function setLocale(string $locale): void
```
didn't we discuss to use a separate service for this, to prevent the framework from having public methods that are not API? A separate service could have `@internal`.

fritzmg replied:
> didn't we discuss to use a separate service for this, to prevent the framework from having public methods that are not API?

See https://github.com/contao/contao/pull/8691#issuecomment-3211753354 - a separate service seemed superfluous to me.

aschempp replied:
well I kinda disagree. I use the `ContaoFramework` class in 1000 places, and I wouldn't want to have a `$this->framework->setLocale()` API, because that is internal. And it would most likely not do what people expect. There is no benefit of having this _inside_ the class (apart from having one less class). But it could just as well be in our `LocaleSubscriber`?

fritzmg replied:
> But it could just as well be in our `LocaleSubscriber`?

You mean the `LocaleSubscriber` should implement the `LocaleAwareInterface`?

fritzmg replied:
Just to elaborate: the `ContaoFramework` implementing the `LocaleAwareInterface` made sense to me, since it needs to be aware of the set/changed locale in order to change the legacy `$GLOBALS['TL_LANGUAGE']` global. It already sets it during initialization to the locale of the request - and thus it should also set it, when the locale is changed.

aschempp replied:
It could, right? And it kinda already does stuff with that?

fritzmg replied:
I don't know, I would find it odd to add that interface to the event listener. `ContaoFramework` makes still more sense to me.

fritzmg replied:
Nevertheless, separate service now implemented.

fritzmg committed:
Revert only creating the response context if it does not exist yet

fritzmg committed:
Do not use request locale

fritzmg committed:
fix test

fritzmg committed:
Merge branch '5.6' into bugfix/load-default-translations-layout-page-controller

fritzmg committed:
Merge remote-tracking branch 'origin/5.6' into bugfix/load-default-translations-layout-page-controller

fritzmg committed:
Use LocaleSwitcher in LocaleSubscriber

fritzmg committed:
use separate service

fritzmg committed:
revert comment change

leofeyer committed:
Merge branch '5.6' into bugfix/load-default-translations-layout-page-controller

leofeyer committed:
CS
